### DISCUSSION TITLE
Ability to take address of parameters

### DISCUSSION THREAD

**Discussion by alexover1** [odin-lang/Odin#4312] [Ideas/Requests]
Procedure parameters are currently not allowable to take the pointer address of. This makes sense, because they are not modifiable. However, there are some cases where it would be useful, and could avoid a useless copy.

```odin
// Send light properties to shader
update_light_values :: proc(shader: Shader, light: Light) {
    light := light
    rl.SetShaderValue(shader, light.enabled_loc, &light.enabled, .INT)
    rl.SetShaderValue(shader, light.type_loc, &light.type, .INT)
}
```

What about allowing taking the address of #by_ptr parameters?

```odin
// Send light properties to shader
update_light_values :: proc(shader: Shader, #by_ptr light: Light) {
    rl.SetShaderValue(shader, light.enabled_loc, &light.enabled, .INT)
    rl.SetShaderValue(shader, light.type_loc, &light.type, .INT)
}
```

**Comment by laytan** [comment]
The reason for this is that Odin can choose to pass the parameter by reference, if you allow taking the address of that you would be able to change the value of the caller's version of the value which is not what you want.

We also don't want to automatically do `light := light` for you because it may be undesirable to do for performance, stack usage, or some other reason. We want to explicitly give the writer the choice of doing that or passing by pointer in the first place.
